# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [32]:
import numpy as np # importing numpy for numerical computation
import random
import scipy
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
from sklearn.tree import DecisionTreeRegressor
from prettytable import PrettyTable

In [33]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [34]:
x.shape

(506, 13)

In [35]:
y.shape

(506,)

In [36]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

In [37]:
y[:5]

array([24. , 21.6, 34.7, 33.4, 36.2])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [38]:
def generating_samples(input_data, target_data):
    
    '''In this function, we will write code for generating 30 samples'''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    # note please return as lists
    selecting_rows = np.random.choice(len(input_data), 303, replace=False)
    replacing_rows = np.random.choice(selecting_rows, 203, replace=False)
    number_of_columns_to_select = random.randint(3, 13)
    selecting_columns = np.array(random.sample(range(0, 13), number_of_columns_to_select ))
    sample_data = input_data[selecting_rows[:, None], selecting_columns]
    target_of_sample_data = target_data[selecting_rows]
    
    #replicating data
    replicated_sample_data = input_data[replacing_rows[:, None], selecting_columns]
    target_of_replicated_sample_data = target_data[replacing_rows]
    
    #concatenating data
    final_sample_data = np.vstack((sample_data, replicated_sample_data ))
    final_target_data = np.vstack((target_of_sample_data.reshape(-1, 1), target_of_replicated_sample_data.reshape(-1, 1)))
    
    return final_sample_data, final_target_data, selecting_rows, selecting_columns
  # random.choice to generate random indices without replacement
  # selecting 303 random row indices from the input_data, without replacement
  # now we will replicate 203 points from above selected rows
  # Replacing Rows => Extracting 206 reandom row indices from the abvoe rows_selected
  # Now get 3 to 13 random column indices from input_data
  # Now Replication of Data for 203 data points out of 303 selected points
  # Concatenating data

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [39]:
def grader_samples(a,b,c,d):
        length = (len(a)==506  and len(b)==506)
        sampled = (len(a)-len(set([str(i) for i in a]))==203)
        rows_length = (len(c)==303)
        column_length= (len(d)>=3)
        assert(length and sampled and rows_length and column_length)
        return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [40]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range (0, 30):
    a, b, c, d = generating_samples(x, y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [41]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [42]:
list_of_all_models_decision_tree = []
for i in range(0, 30):
    model_i = DecisionTreeRegressor(max_depth=None)
    model_i.fit(list_input_data[i], list_output_data[i])
    list_of_all_models_decision_tree.append(model_i)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [44]:
array_of_Y = []

for i in range(0, 30):
    data_point_i = x[:, list_selected_columns[i]]
    target_y_i = list_of_all_models_decision_tree[i].predict(data_point_i)
    array_of_Y.append(target_y_i)
  
  
predicted_array_of_target_y = np.array(array_of_Y)
predicted_array_of_target_y = predicted_array_of_target_y.transpose()

# print(predicted_array_of_target_y.shape)

# Now to calculate MSE, first calculate the Median of Predicted Y
# passing axis=1 will make sure the medians are computed along axis=1
median_predicted_y = np.median(predicted_array_of_target_y, axis=1)
median_predicted_y.shape

print("MSE : ", mean_squared_error(y, median_predicted_y ))

MSE :  0.006551383399209485


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [46]:
y_predicted_oob_median_list=[]
for i in range(0,506):
    
    indicies_for_oob_models=[]
    for index_oob in range(0,30):
        if i not in list_selected_row[index_oob]:
            indicies_for_oob_models.append(index_oob)
    y_predicted_oob_list = []
    for oob_model_index in indicies_for_oob_models:
        model_oob=list_of_all_models_decision_tree[oob_model_index]
        row_oob=x[i]
        x_oob_data_point=[row_oob[columns] for columns in list_selected_columns[oob_model_index]]
        x_oob_data_point=np.array(x_oob_data_point).reshape(1,-1)
        y_predicted_oob_data_point=model_oob.predict(x_oob_data_point)
        y_predicted_oob_list.append(y_predicted_oob_data_point)
    y_predicted_oob_list=np.array(y_predicted_oob_list)
    y_predicted_median=np.median(y_predicted_oob_list)
    y_predicted_oob_median_list.append(y_predicted_median)

def caluclate_oob_score(num_rows):
    oob_score=0
    for i in range(0,num_rows):
        oob_score+=((y[i]-y_predicted_oob_median_list[i])**2)
    final_oob_score=oob_score/506
    return final_oob_score
print("final_oob_score is",caluclate_oob_score(506))

final_oob_score is 14.850484189723307


# <font color='blue'><b>Task 2</b></font>

In [86]:
def bootstrapping_and_oob(x, y):

  # Use generating_samples function to create 30 samples 
  # store these created samples in a list
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]
  
    for i in range (0, 30):
        a, b, c, d = generating_samples(x, y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
  
  # building regression trees
    list_of_all_models_decision_tree = []
    for i in range(0, 30):
        model_i = DecisionTreeRegressor(max_depth=None)
        model_i.fit(list_input_data[i], list_output_data[i])
        list_of_all_models_decision_tree.append(model_i)
  
  # calculating MSE
    array_of_Y = []

    for i in range(0, 30):
        data_point_i = x[:, list_selected_columns[i]]
        target_y_i = list_of_all_models_decision_tree[i].predict(data_point_i)
        array_of_Y.append(target_y_i)
    
    
    predicted_array_of_target_y = np.array(array_of_Y)
    predicted_array_of_target_y = predicted_array_of_target_y.transpose()

  # print(predicted_array_of_target_y.shape)

  # Now to calculate MSE, first calculate the Median of Predicted Y
  # passing axis=1 will make sure the medians are computed along axis=1
    median_predicted_y = np.median(predicted_array_of_target_y, axis=1)
  
  # And now the final MSE
    MSE = mean_squared_error(y, median_predicted_y )
  
  # Calculating OOB Score
    y_predicted_oob_median_list = []

    for i in range(0, 506):
        indices_for_oob_models = []
    
    # For each of i-th row I shall build a list of sample size 30
    # ONLY condition being that this ith row should not be part of
    # the list_selected_row
        for index_oob in range(0, 30):
            if i not in list_selected_row[index_oob]:
                indices_for_oob_models.append(index_oob)
        
        y_predicted_oob_list = []
    
        for oob_model_index in indices_for_oob_models:
            model_oob = list_of_all_models_decision_tree[oob_model_index]
      
            row_oob = x[i]
      # print('oob_model_index ', oob_model_index)
      
            x_oob_data_point = [row_oob[col] for col in list_selected_columns[oob_model_index] ]
      # print('np.array(x_oob_data_point) ', np.array(x_oob_data_point))
            x_oob_data_point = np.array(x_oob_data_point).reshape(1, -1)
      
            y_predicted_oob_data_point = model_oob.predict(x_oob_data_point)
            y_predicted_oob_list.append(y_predicted_oob_data_point)
      # 
        y_predicted_oob_list = np.array(y_predicted_oob_list)
    
        y_predicted_median = np.median(y_predicted_oob_list)
        y_predicted_oob_median_list.append(y_predicted_median)
    

    oob_score = 0

    for i in range(0, 506):
        oob_score += (y[i] - y_predicted_oob_median_list[i] ) ** 2
    # oob_score = (oob_score + (y[i] - y_predicted_oob_median_list[i] ) ** 2)
    # 13.828377285079045
        

    final_oob_score = oob_score/506
  
    return MSE, final_oob_score

print(bootstrapping_and_oob(x, y))

(0.05208027744290138, 15.425342186186965)


In [87]:
x=boston.data #independent variables
y=boston.target #target variable

mse_boston_35_times_arr = []
oob_score_boston_35_times_arr = []

# Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score
for i in range(0, 35):
    mse, oob_score = bootstrapping_and_oob(x, y)
    mse_boston_35_times_arr.append(mse)
    oob_score_boston_35_times_arr.append(oob_score)

mse_boston_35_times_arr = np.array(mse_boston_35_times_arr)
oob_score_boston_35_times_arr = np.array(oob_score_boston_35_times_arr)

confidence_level = 0.95
degrees_of_freedom = 34 # sample.size - 1

mean_of_sample_mse_35 = np.mean(mse_boston_35_times_arr)
standard_error_of_sample_mse_35 = scipy.stats.sem(mse_boston_35_times_arr)


# Per document - https://www.kite.com/python/answers/how-to-compute-the-confidence-interval-of-a-sample-statistic-in-python
# confidence_interval = scipy.stats.t.interval(confidence_level, degrees_freedom, sample_mean, sample_standard_error)
confidence_interval_mse_35 = scipy.stats.t.interval(confidence_level, degrees_of_freedom, mean_of_sample_mse_35, standard_error_of_sample_mse_35 )
print("confidence_interval_mse_35 ", confidence_interval_mse_35)

# Now calculate confidence inter for oob score
mean_of_sample_oob_score_35 = np.mean(oob_score_boston_35_times_arr)
standard_error_of_sample_oob_score_35 = scipy.stats.sem(oob_score_boston_35_times_arr)

confidence_interval_oob_score_35 = scipy.stats.t.interval(confidence_level, degrees_of_freedom, mean_of_sample_oob_score_35, standard_error_of_sample_oob_score_35 )
print("confidence_interval_oob_score_35 ", confidence_interval_oob_score_35)

confidence_interval_mse_35  (0.07776336237970008, 0.13612551351201704)
confidence_interval_oob_score_35  (13.298451267890044, 14.702814615706075)


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [47]:
def predict_y_given_x(x_query):
    y_predicted_array_30_sample = []

    for i in range(0, 30):
        model_i = list_of_all_models_decision_tree[i]
    # Extract x for ith data point with specific number of featues from list_selected_columns
        x_data_point_i = [x_query[column] for column in list_selected_columns[i]]
        x_data_point_i = np.array(x_data_point_i).reshape(1, -1)
        y_predicted_i = model_i.predict(x_data_point_i)
        y_predicted_array_30_sample.append(y_predicted_i)

    y_predicted_array_30_sample = np.array(y_predicted_array_30_sample)
    y_predicted_median = np.median(y_predicted_array_30_sample)
    return y_predicted_median

xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]
y_predicted_for_xq = predict_y_given_x(xq)
y_predicted_for_xq

18.5

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

By definition we know the interpretation of a 95% confidence interval for the populace mean as - If rehashed random samples  were taken and the 95% confidence interval was figured for each example, 95% of the intervals  would contain the population mean.

MSE - 
There is a 95% chance that the confidence interval of (0.07776336237970008, 0.13612551351201704) contains the true population mean of MSE.

OOB Score - 
There is a 95% chance that the confidence interval of (13.298451267890044, 14.702814615706075) contains the true population mean of OOB Score.
